### ISSUE TITLE
Missing docs in changes feed after auto failover (channel-cache not updating)

### ISSUE THREAD

**Issue by sethrosetter** [couchbase/sync_gateway#2197]
### Sync Gateway version
Couchbase Sync Gateway/1.4(12;d19597f) Channel Cache mode

### Operating system
CentOS 7

### Config file
standard

### Log output
https://gist.github.com/sethrosetter/12bd611d31cef9e7f9193ef815124bd7

### Expected behavior
Using a 3-node Couchbase Server cluster, enable auto failover. After taking a server down (via `service couchbase-server stop`), start adding docs in a loop (100 docs each iteration). When the docs succeed again (after 30 second failover pause), exit the loop but maintain the ids of the docs that were added. 

Verify that the all the docs were added (True)
Verify that all the docs show up in the changes feed (True)

### Actual behavior
Verify that all the docs show up in the changes feed (False)

### Steps to reproduce
1.  Run https://github.com/couchbaselabs/mobile-testkit/blob/feature/rebalance-support/testsuites/syncgateway/functional/topology_specific_tests/multiple_servers/test_muliple_servers.py#L92-L191

### Interesting note!!
It seems as though the channel cache gets into a corrupt state due to the failover. When restarting the sync_gateway, all of the changes show up in the changes feed.

**Comment by adamcfraser** [comment]
Are your expected and actual switched above?  (False/True)

This sounds like a case of Sync Gateway dropping the Tap feed during failover - possibly related to #2153.

**Comment by tleyden** [comment]
@sethrosetter were you able to repro with distributed index mode?

**Comment by sethrosetter** [comment]
@adamcfraser sorry. Mixed those up. Fixed. 
@tleyden I have not. I will run today.

**Comment by sethrosetter** [comment]
The test fails with distributed index as well.

**Comment by sethrosetter** [comment]
We need to reenable this before closing - https://github.com/couchbaselabs/mobile-testkit/blob/master/testsuites/syncgateway/functional/topology_specific_tests/multiple_servers/test_multiple_servers.py#L94-L200

**Comment by djpongh** [comment]
@sethrosetter can you re-enable the above test and retest?

**Comment by raghusarangapani** [comment]
Just ran test_multiple_servers.py without xattrs, in di mode:

All the 3 tests failed with the error `E           AssertionError: Failed to start sg_accel`

Command:
```
pytest -s --mode=di --server-version=5.0.0-3456 --sync-gateway-version=1.5.0-515 testsuites/syncgateway/functional/topology_specific_tests/multiple_servers/test_multiple_servers.py
```

**Comment by raghusarangapani** [comment]
Ran this test in cc mode: test_server_goes_down_sanity consistently fails with `E           TimeoutException: Failed to get all changes`

Command:
```
pytest -s -k test_server_goes_down_sanity --mode=cc --server-version=5.0.0-3456 --sync-gateway-version=1.5.0-515 testsuites/syncgateway/functional/topology_specific_tests/multiple_servers/test_multiple_servers.py
```

**Comment by tleyden** [comment]
Fixed by https://github.com/couchbaselabs/sync-gateway-accel/pull/171

Pending re-verification, but requires https://github.com/couchbaselabs/mobile-testkit/issues/1338

**Comment by sridevi-15** [comment]
I added support of multiple support in framework and ran the test , but failing to get changes.

sync-gateway config looks like this : 

```
ok: [sg1] => {
    "out.stdout": {
        "adminInterface": "0.0.0.0:4985", 
        "cluster_config": {
            "bucket": "data-bucket", 
            "data_dir": ".", 
            "password": "password", 
            "server": "http://192.168.33.10,192.168.33.11,192.168.33.12:8091", 
            "username": "data-bucket"
        }, 
        "compressResponses": false, 
        "databases": {
            "db": {
                "bucket": "data-bucket", 
                "channel_index": {
                    "bucket": "index-bucket", 
                    "num_shards": 16, 
                    "password": "password", 
                    "server": "http://192.168.33.10,192.168.33.11,192.168.33.12:8091", 
                    "username": "index-bucket", 
                    "writer": false
                }, 
                "password": "password", 
                "server": "http://192.168.33.10,192.168.33.11,192.168.33.12:8091", 
                "username": "data-bucket"
            }
        }, 
        "interface": ":4984", 
        "log": [
            "CRUD+", 
            "Cache+", 
            "HTTP+", 
            "Changes+"
        ], 
        "maxCouchbaseConnections": 16, 
        "maxFileDescriptors": 90000, 
        "maxIncomingConnections": 0, 
        "slowServerCallWarningThreshold": 500
    }
}

```

Btw, test_rebalance_sanity passed with multiple url support

**Comment by adamcfraser** [comment]
@sridevi-15 If this wasn't a jenkins job (and so can't link to the failing test), it would be useful to get the Sync Gateway logs included on the ticket so that we can see exactly how it's failing to get the changes.

**Comment by sridevi-15** [comment]
@adamcfraser , here is the jenkins link : http://uberjenkins.sc.couchbase.com/job/cen7-sync-gateway-functional-tests-topology-specific-di/715/testReport/

logs are in this link under build artifacts : http://uberjenkins.sc.couchbase.com/job/cen7-sync-gateway-functional-tests-topology-specific-di/715/

**Comment by adamcfraser** [comment]
Based on the accel logs, it looks to me like accel is still getting connection, rollback and refresh worker errors for up to 5 minutes after the server is brought down.  

There are several repeated rollback notifications for the same vbucket, which suggests that accel isn't reconnecting to the stream property.

@sridevi-15 A couple of requests to help debug this further:

1. Can this test be rerun with log `*` enabled on sg-accel?  That will provide more visibility into what's going on.
2. What's the status of this test when running in channel cache mode?

**Comment by sridevi-15** [comment]
Here is the sg-config : 
```
ok: [sg1] => {
    "out.stdout": {
        "adminInterface": "0.0.0.0:4985", 
        "compressResponses": false, 
        "databases": {
            "db": {
                "bucket": "data-bucket", 
                "password": "password", 
                "server": "http://192.168.33.10,192.168.33.11,192.168.33.12:8091", 
                "username": "data-bucket"
            }
        }, 
        "interface": ":4984", 
        "log": [
            "*"
        ], 
        "maxCouchbaseConnections": 16, 
        "maxFileDescriptors": 90000, 
        "maxIncomingConnections": 0, 
        "slowServerCallWarningThreshold": 500
    }
}

```

Ran with channel cache mode and see same issue. I enabled log * in sg. 
I tried increasing the timeout to 5 minutes and I manually checked _changes and I see 36 missing docs which is what test is also reported. Out of 100 docs, it found 64 and missing 36 docs.


Here is the sg logs attached 


[sync_gateway_error.zip](https://github.com/couchbase/sync_gateway/files/1268045/sync_gateway_error.zip)

I see this error for missing docs in sg logs: One of the missing doc is 84
2017-08-31T04:08:20.550Z Cache+:   Ignoring duplicate of #84
2017-08-31T04:08:20.916Z WARNING: DCP RollbackEx request - rolling back DCP feed for: vbucketId: 437, rollbackSeq: 0 -- base.(*DCPReceiver).RollbackEx() at dcp_feed.go:259

**Comment by tleyden** [comment]
Reproduced locally on vagrant cluster with:

```
pytest -s -k test_server_goes_down_sanity --mode=cc --server-version=5.0.0-3456 --sync-gateway-version=1.5.0-555 testsuites/syncgateway/functional/topology_specific_tests/multiple_servers/test_multiple_servers.py
```

Mobile-testkit logs: https://gist.github.com/tleyden/24fb3ca7858faa2bbf673644e194dbd7
SG logs: https://gist.github.com/tleyden/baec2a7161feb1db2e9d56ebd1ff0044

**Comment by tleyden** [comment]
With slightly modified mobile-testkit commit: 836a8e51e05601bc01e6b92b14bdec2a0b25605c (pauses after detects changes feed verification problem), I'm seeing these logs:


https://gist.github.com/tleyden/43f2cb0519a91fa97f3dcf876bcae965

Note that that one of the missing docs in the _changes feed:

```
2017-09-01 15:05:00.846865 Example missing expected doc id: 492b5755-8c0d-48a8-9b4d-f8241ffb8c80
```

Never seems to come in a `DataUpdate` DCP mutation:

```
[root@localhost logs]# grep -i 492b5755-8c0d-48a8-9b4d-f8241ffb8c80 sync_gateway_error.log
2017-09-01T22:03:05.936Z HTTP:  #062: PUT /db/492b5755-8c0d-48a8-9b4d-f8241ffb8c80  (as seth)
2017-09-01T22:03:05.936Z CRUD+: Invoking sync on doc "492b5755-8c0d-48a8-9b4d-f8241ffb8c80" rev 1-4fe20cb80892e788f7294dee7f79e347
2017-09-01T22:03:05.937Z CRUD: 	Doc "492b5755-8c0d-48a8-9b4d-f8241ffb8c80" in channels "{ABC, CBS}"
2017-09-01T22:03:05.937Z CRUD+: Saving doc (seq: #61, id: 492b5755-8c0d-48a8-9b4d-f8241ffb8c80 rev: 1-4fe20cb80892e788f7294dee7f79e347)
2017-09-01T22:03:05.937Z CRUD: Stored doc "492b5755-8c0d-48a8-9b4d-f8241ffb8c80" / "1-4fe20cb80892e788f7294dee7f79e347"
```

**Comment by tleyden** [comment]
There is a possible go-couchbase / cbdatasource issue.  I filed https://issues.couchbase.com/browse/MB-25912 to try to get to the bottom of that, or find out how we should be calling cbdatasource differently.

**Comment by tleyden** [comment]
Tested the proposed fix for MB-25912, and it looks like the test is passing now.

Mobile-testkit + SG Logs: https://gist.github.com/tleyden/864c8847f08231df77cbd09984dd4ecf

Once MB-25912 is merged into master, we can uptake the latest go-couchbase to get the fix

**Comment by sridevi-15** [comment]
Verified on 1.5.0-573 and working fine.
